File: 2025/day_05/cafeteria.py
#!/usr/bin/python3

import sys
from functools import reduce


def main(filename):
    with open(filename, "r") as file:
        content = file.read()

    parts = content.strip().split("\n\n")
    integers = list(map(int, parts[1].split("\n")))
    ranges = [
        range(int(n), int(m) + 1)
        for n, m in [line.split("-") for line in parts[0].split("\n")]
    ]

    return (
        sum(1 for id in integers if any(id in r for r in ranges)),
        sum(len(r) for r in merge_overlapping_ranges(ranges)),
    )


def merge_overlapping_ranges(ranges):
    if not ranges:
        return []

    sorted_ranges = sorted(ranges, key=lambda r: r.start)
    return reduce(merge_if_overlapping, sorted_ranges, [])


def merge_if_overlapping(merged, current):
    if not merged:
        return [current]

    last = merged[-1]
    if last.stop <= current.start:
        return merged + [current]

    new_start = min(last.start, current.start)
    new_stop = max(last.stop, current.stop)
    return merged[:-1] + [range(new_start, new_stop)]


if __name__ == "__main__":
    part_1, part_2 = main(sys.argv[1])
    print(f"Part 1:\t{part_1}")
    print(f"Part 2:\t{part_2}")

File: 2025/day_05/cafeteria.exs
#!/usr/bin/env elixir

defmodule IngredientProcessor do
  def main(filename) do
    {:ok, content} = File.read(filename)

    parts = String.split(content, "\n\n", trim: true)

    integers = parts
    |> Enum.at(1)
    |> String.split("\n", trim: true)
    |> Enum.map(&String.to_integer/1)

    ranges = parts
    |> Enum.at(0)
    |> String.split("\n", trim: true)
    |> Enum.map(fn line ->
      [n, m] = String.split(line, "-", trim: true) |> Enum.map(&String.to_integer/1)
      n..(m)
    end)

    Enum.count(integers, fn id ->
      Enum.any?(ranges, fn range -> id in range end)
    end)
    |> IO.inspect(label: "Part 1")

    ranges
    |> merge_overlapping_ranges()
    |> Enum.map(fn range -> range.last - range.first + 1 end)
    |> Enum.sum
    |> IO.inspect(label: "Part 2")
  end

  def merge_overlapping_ranges(ranges) do
    case ranges do
      [] -> []
      _ ->
        sorted_ranges = Enum.sort(ranges, fn r1, r2 -> r1.first < r2.first end)
        reduce_ranges(sorted_ranges, [])
    end
  end

  defp reduce_ranges([], acc), do: Enum.reverse(acc)

  defp reduce_ranges([current | rest], acc) do
    case acc do
      [] -> reduce_ranges(rest, [current])
      [last | _] ->
        if last.last < current.first do
          reduce_ranges(rest, [current | acc])
        else
          new_start = min(last.first, current.first)
          new_stop = max(last.last, current.last)
          reduce_ranges(rest, [new_start..new_stop | tl(acc)])
        end
    end
  end
end

case System.argv() do
  [filename] -> IngredientProcessor.main(filename)
  _ -> IO.puts("Usage: elixir ingredient_processor.exs <filename>")
end
